src/Nethermind/Nethermind.State/WorldState.cs

```diff
@@ -248,32 +316,106 @@ public Snapshot TakeSnapshot(bool newTransactionStart = false)
 
         public void Restore(Snapshot snapshot)
         {
+            EnsureInScope();
             _persistentStorageProvider.Restore(snapshot.StorageSnapshot.PersistentStorageSnapshot);
             _transientStorageProvider.Restore(snapshot.StorageSnapshot.TransientStorageSnapshot);
             _stateProvider.Restore(snapshot.StateSnapshot);
         }
 
         internal void Restore(int state, int persistentStorage, int transientStorage)
         {
+            EnsureInScope();
             Restore(new Snapshot(state, new Snapshot.Storage(persistentStorage, transientStorage)));
         }
 
         public void SetNonce(Address address, in UInt256 nonce)
         {
+            EnsureInScope();
             _stateProvider.SetNonce(address, nonce);
         }
 
         public void CreateAccountIfNotExists(Address address, in UInt256 balance, in UInt256 nonce = default)
         {
+            EnsureInScope();
             _stateProvider.CreateAccountIfNotExists(address, balance, nonce);
         }
 
-        ArrayPoolList<AddressAsKey>? IWorldState.GetAccountChanges() => _stateProvider.ChangedAddresses();
+        ArrayPoolList<AddressAsKey>? IWorldState.GetAccountChanges()
+        {
+            EnsureInScope();
+            return _stateProvider.ChangedAddresses();
+        }
+
         public void ResetTransient()
         {
+            EnsureInScope();
             _transientStorageProvider.Reset();
         }
 
         PreBlockCaches? IPreBlockCaches.Caches => PreBlockCaches;
+
+        private void EnsureInScope()
+        {
+            if (!_inScope) throw new InvalidOperationException("WorldState used outside of its scope.");
+        }
+
+        private bool _inScope;
+
+        public IDisposable BeginScope(Hash256 stateRoot)
+        {
+            if (_inScope)
```

Comment:
Do this case happen? Is this expected?

Reply:
I'd capture the current scope as a field instead of boolean and bump up the counter there. Otherwise you can run into situations where nested scopes are not disposed and you won't get any notification about it. With a counter you get all the book keeping you need.

Reply:
Ok. One more thing. Why don't we capture the previous root hash and reset to it at the end of the scope? That would be a better scoping mechanism?